pkg/server/api/middleware/authorization.go

```diff
@@ -59,18 +51,18 @@ func (m *authorizationMiddleware) Preprocess(ctx context.Context, methodName str
 		ctx = rpccontext.WithLogger(ctx, rpccontext.Logger(ctx).WithFields(fields))
 	}
 
-	authorizer, ok := m.authorizers[methodName]
-	if !ok {
-		middleware.LogMisconfiguration(ctx, "Authorization misconfigured (method not registered); this is a bug")
-		return nil, status.Errorf(codes.Internal, "authorization misconfigured for %q (method not registered)", methodName)
-	}
-
-	authorizedCtx, err := authorizer.AuthorizeCaller(ctx)
+	ctx, allow, err := m.opaAuth(ctx, req, methodName)
 	if err != nil {
 		rpccontext.Logger(ctx).WithError(err).Error("Failed to authenticate caller")
 		return nil, err
 	}
-	return authorizedCtx, nil
+	if allow {
+		return ctx, nil
+	}
+
+	deniedErr := status.Errorf(codes.PermissionDenied, "Authorization denied for method %s", methodName)
```

Comment:
This entire effort is a fantastic security addition. Has there been discussion for future work on enabling triage of auth failures? In particular, I am thinking of the Batch APIs. On rejection, the caller (and SPIRE operators) only seem to know that at least one of the entries in the Batch was denied, but not which one. Depending on complexity of policies being set, it could be a bit of work to determine denial reason. If the caller was a workload, it is also cumbersome to get the entire list of entries printed out.

Reply:
yea, that's a great idea!!! :). Having error messages being propagated from the OPA engine would be super useful. I'm still unsure what the best way to do this is with OPA. Let me ask some of the OPA folks to see if they have anything that can do this in the pipeline or if they have some ideas of how this can be done today.

Reply:
In general, reason of denial is a pretty big deal, and as complexity grows this gets more important as @amoore877 is noting here. This is another general issue with adopting opa/rego in this fashion, so if there's not an existing solution I imagine that there is a community need for one.

Reply:
There's an example of it here https://play.openpolicyagent.org/p/giimmNjfrj

it does look like the policies need to be instantiated a certain way - like a deny list rather than an allow list , will have to look into this a bit more.